src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/rewriting/RuleApplicationBaseImpl.java

```diff
@@ -0,0 +1,75 @@
+package se.liu.ida.hefquin.engine.queryproc.impl.optimizer.rewriting;
+
+import se.liu.ida.hefquin.engine.queryplan.PhysicalPlan;
+import se.liu.ida.hefquin.engine.queryplan.physical.BinaryPhysicalOp;
+import se.liu.ida.hefquin.engine.queryplan.physical.NaryPhysicalOp;
+import se.liu.ida.hefquin.engine.queryplan.physical.UnaryPhysicalOp;
+import se.liu.ida.hefquin.engine.queryplan.physical.impl.PhysicalPlanWithBinaryRootImpl;
+import se.liu.ida.hefquin.engine.queryplan.physical.impl.PhysicalPlanWithNaryRootImpl;
+import se.liu.ida.hefquin.engine.queryplan.physical.impl.PhysicalPlanWithUnaryRootImpl;
+
+public abstract class RuleApplicationBaseImpl implements RuleApplication{
+    protected final PhysicalPlan[] subPlans;
+    protected RewritingRule rule;
+
+    public RuleApplicationBaseImpl( final PhysicalPlan[] subPlans, final RewritingRule rule ){
+        this.subPlans = subPlans;
+        this.rule = rule;
+    }
+
+    @Override
+    public PhysicalPlan getPlan() {
+        return subPlans[0];
+    }
+
+    @Override
+    public RewritingRule getRule() {
+        return rule;
+    }
+
+    @Override
+    public PhysicalPlan getResultingPlan() {
+        final int numStep = subPlans.length;
+        PhysicalPlan rewrittenPlan = rewrittenSubPlan( subPlans[numStep-1] );
+
+        for (int i = numStep-2; i >= 0; i--) {
```

Comment:
Since the `numStep` variable is _not_ meant to represent the number of a step but, instead, the total number of step**s**, it should be called `numSteps`. Also, you can immediately initialize it to `subPlans.length - 1`.